src/requests/sign-request.ts

```diff
@@ -0,0 +1,125 @@
+import * as crypto from 'crypto'
+import {
+  CanonicalRequest,
+  NormalizedCanonicalRequest,
+  Secret,
+  SignedRequestHeaders,
+  Timestamp,
+  ContentfulHeader,
+} from './typings'
+import { CanonicalRequestValidator, SecretValidator, TimestampValidator } from './typings'
+import { getNormalizedEncodedURI, normalizeHeaders, sortHeaderKeys } from './utils'
+
+const hash = (normalizedCanonicalRequest: NormalizedCanonicalRequest, secret: string) => {
+  const stringifiedHeaders = normalizedCanonicalRequest
+    .headers!.map(([key, value]) => `${key}:${value}`)
+    .join(';')
+
+  const stringifiedRequest = [
+    normalizedCanonicalRequest.method,
+    normalizedCanonicalRequest.path,
+    stringifiedHeaders,
+    normalizedCanonicalRequest.body,
+  ].join('\n')
+
+  const hmac = crypto.createHmac('sha256', secret)
+
+  hmac.update(stringifiedRequest)
+
+  return hmac.digest('hex')
+}
+
+const getSortedAndSignedHeaders = (headers: Record<string, string>, timestamp: number) => {
+  const rawSignedHeaders = Object.keys(headers)
+
+  if (!(ContentfulHeader.SignedHeaders in headers)) {
+    rawSignedHeaders.push(ContentfulHeader.SignedHeaders)
+  }
+
+  if (!(ContentfulHeader.Timestamp in headers)) {
+    rawSignedHeaders.push(ContentfulHeader.Timestamp)
+  }
+
+  const signedHeaders = rawSignedHeaders.sort(sortHeaderKeys)
+
+  const signedHeadersString = signedHeaders.join(',')
+
+  headers[ContentfulHeader.Timestamp] = timestamp.toString()
+  headers[ContentfulHeader.SignedHeaders] = signedHeadersString
+
+  const sortedHeaders = Object.entries(headers).sort(([keyA], [keyB]) => sortHeaderKeys(keyA, keyB))
+
+  return { sortedHeaders, signedHeaders: signedHeadersString }
+}
+
+/**
+ * Given a secret, a canonical request and a timestamp, generates a signature.
+ * It can be used to verify canonical requests to assess authenticity of the
+ * sender and integrity of the payload.
+ *
+ * ~~~
+ * const {signRequest, ContentfulHeader} = require('contentful-node-apps-toolkit')
+ * const {pick} = require('lodash')
+ * const {server} = require('./imaginary-server')
+ *
+ * const SECRET = process.env.SECRET
+ *
+ * server.post('/api/my-resources', (req, res) => {
+ *   const incomingSignature = req.headers['x-contentful-signature']
+ *   const incomingTimestamp = req.headers['x-contentful-timestamp']
+ *   const incomingSignedHeaders = req.headers['x-contentful-signed-headers']
+ *   const now = Date.now()
+ *
+ *   if (!incomingSignature) {
+ *     res.send(400, 'Missing signature')
+ *   }
+ *
+ *   if (now - incomingTimestamp > 1000) {
+ *     res.send(408, 'Request too old')
+ *   }
+ *
+ *   const signedHeaders = incomingSignedHeaders.split(',')
+ *
+ *   const {[ContentfulHeader.Signature]: computedSignature} = signRequest(
+ *     SECRET,
+ *     {
+ *       method: req.method,
+ *       path: req.url,
+ *       headers: pick(req.headers, signedHeaders),
+ *       body: JSON.stringify(req.body)
+ *     },
+ *     incomingTimestamp
```

Comment:
This actually doesn't work, because `incomingTimestamp` is a string, and a number is expected. We should add a typecast to this.